// valid.js/transpiler/validate.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}(g.valid || (g.valid = {})).js = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
"use strict";
Object.defineProperty(exports, "__esModule", {value: true});
var isTrue = function isTrue(value) {
  return value === true;
};
var isFalse = function isFalse(value) {
  return value === false;
};
var bool = exports.bool = {
  isTrue: isTrue,
  isFalse: isFalse
};

},{}],2:[function(require,module,exports){
"use strict";
"use strict";
Object.defineProperty(exports, "__esModule", {value: true});
var isVisa = function isVisa(creditCard) {
  var visaRegex = /^4[0-9]{12}(?:[0-9]{3})?$/;
  return visaRegex.test(creditCard);
};
var isElo = function isElo(creditCard) {
  var eloRegex = /^(401178|401179|431274|438935|451416|457393|457631|457632|504175|627780|636297|636368|(506699|5067[0-6]\d|50677[0-8])|(50900\d|5090[1-9]\d|509[1-9]\d{2})|65003[1-3]|(65003[5-9]|65004\d|65005[0-1])|(65040[5-9]|6504[1-3]\d)|(65048[5-9]|65049\d|6505[0-2]\d|65053[0-8])|(65054[1-9]|6505[5-8]\d|65059[0-8])|(65070\d|65071[0-8])|65072[0-7]|(65090[1-9]|65091\d|650920)|(65165[2-9]|6516[6-7]\d)|(65500\d|65501\d)|(65502[1-9]|6550[3-4]\d|65505[0-8]))[0-9]{10,12}/;
  return eloRegex.test(creditCard);
};
var isHipercard = function isHipercard(creditCard) {
  var hiperCard = /^(38[0-9]{17}|60[0-9]{14})$/;
  return hiperCard.test(creditCard);
};
var isMasterCard = function isMasterCard(creditCard) {
  var masterCardRegex = /^5[1-5][0-9]{14}$/;
  return masterCardRegex.test(creditCard);
};
var isDiners = function isDiners(creditCard) {
  var dinersRegex = /^3(?:0[0-5]|[68][0-9])[0-9]{11}$/;
  return dinersRegex.test(creditCard);
};
var isAmex = function isAmex(creditCard) {
  var amexRegex = /^3[47][0-9]{13}$/;
  return amexRegex.test(creditCard);
};
var isCreditCard = function isCreditCard(creditCard) {
  return isAmex(creditCard) || isDiners(creditCard) || isMasterCard(creditCard) || isHipercard(creditCard) || isElo(creditCard) || isVisa(creditCard) || false;
};
var creditCard = exports.creditCard = {
  isVisa: isVisa,
  isElo: isElo,
  isHipercard: isHipercard,
  isMasterCard: isMasterCard,
  isDiners: isDiners,
  isAmex: isAmex,
  isCreditCard: isCreditCard
};

},{}],3:[function(require,module,exports){
"use strict";
'use strict';
Object.defineProperty(exports, "__esModule", {value: true});
exports.date = undefined;
var _date = require('./lib/date');
var isDate = function isDate(date) {
  return (0, _date._isDate)(date);
};
var maxDate = function maxDate() {
  var max = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : new Date(3000, 1, 1);
  return function(value) {
    return (0, _date._isDate)(value) && value <= max;
  };
};
var minDate = function minDate() {
  var min = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : new Date(1900, 1, 1);
  return function(value) {
    return (0, _date._isDate)(value) && value >= min;
  };
};
var between = function between() {
  var min = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : new Date(1900, 1, 1);
  var max = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : new Date(3000, 1, 1);
  return function(value) {
    return minDate(min)(value) && maxDate(max)(value);
  };
};
var date = exports.date = {
  maxDate: maxDate,
  minDate: minDate,
  between: between,
  isDate: isDate
};

},{"./lib/date":6}],4:[function(require,module,exports){
"use strict";
"use strict";
Object.defineProperty(exports, "__esModule", {value: true});
var isIpv4 = function isIpv4(value) {
  var ipRegex = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
  return ipRegex.test(value);
};
var isIpv6 = function isIpv6(value) {
  var ipRegex = /^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$|^(([a-zA-Z]|[a-zA-Z][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z]|[A-Za-z][A-Za-z0-9\-]*[A-Za-z0-9])$|^\s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:)))(%.+)?\s*$/;
  return ipRegex.test(value);
};
var isIp = function isIp(value) {
  return isIpv4(value) || isIpv6(value);
};
var ip = exports.ip = {
  isIp: isIp,
  isIpv4: isIpv4,
  isIpv6: isIpv6
};

},{}],5:[function(require,module,exports){
"use strict";
'use strict';
Object.defineProperty(exports, "__esModule", {value: true});
var _isArray = exports._isArray = function _isArray(value) {
  return value !== null && value.length >= 0 && Object.prototype.toString.call(value) === '[object Array]';
};

},{}],6:[function(require,module,exports){
"use strict";
'use strict';
Object.defineProperty(exports, "__esModule", {value: true});
var _isDate = exports._isDate = function _isDate(value) {
  return Object.prototype.toString.call(value) === '[object Date]';
};

},{}],7:[function(require,module,exports){
"use strict";
"use strict";
Object.defineProperty(exports, "__esModule", {value: true});
var isNil = exports.isNil = function isNil(value) {
  return value === null || value === undefined;
};

},{}],8:[function(require,module,exports){
"use strict";
'use strict';
Object.defineProperty(exports, "__esModule", {value: true});
var _isNumber = exports._isNumber = function _isNumber(value) {
  return Object.prototype.toString.call(value) === '[object Number]';
};

},{}],9:[function(require,module,exports){
"use strict";
'use strict';
Object.defineProperty(exports, "__esModule", {value: true});
var _isObject = exports._isObject = function _isObject(value) {
  return Object.prototype.toString.call(value) === '[object Object]';
};

},{}],10:[function(require,module,exports){
"use strict";
"use strict";
Object.defineProperty(exports, "__esModule", {value: true});
var reduce = exports.reduce = function reduce(func, index, array) {
  if (Array.prototype.reduce) {
    return Array.prototype.reduce.apply(array, [func, index]);
  }
  var iterator = array[Symbol.iterator]();
  var go = function go(acc, iterable) {
    var _iterable$next = iterable.next();
    var value = _iterable$next.value;
    var done = _iterable$next.done;
    return done ? acc : go(func(acc, value), iterable);
  };
  return go(index, iterator);
};

},{}],11:[function(require,module,exports){
"use strict";
'use strict';
Object.defineProperty(exports, "__esModule", {value: true});
var _isString = exports._isString = function _isString(value) {
  return Object.prototype.toString.call(value) === '[object String]';
};

},{}],12:[function(require,module,exports){
"use strict";
'use strict';
Object.defineProperty(exports, "__esModule", {value: true});
exports.number = undefined;
var _number = require('./lib/number');
var isNumber = function isNumber(number) {
  return (0, _number._isNumber)(number);
};
var maxNumber = function maxNumber() {
  var maxValue = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : Number.MAX_VALUE;
  return function(value) {
    return (0, _number._isNumber)(value) && value <= maxValue;
  };
};
var minNumber = function minNumber() {
  var minValue = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : Number.MIN_VALUE;
  return function(value) {
    return (0, _number._isNumber)(value) && value >= minValue;
  };
};
var between = function between() {
  var minValue = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : Number.MIN_VALUE;
  var maxValue = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : Number.MAX_VALUE;
  return function(value) {
    return minNumber(minValue)(value) && maxNumber(maxValue)(value);
  };
};
var number = exports.number = {
  maxNumber: maxNumber,
  minNumber: minNumber,
  between: between,
  isNumber: isNumber
};

},{"./lib/number":8}],13:[function(require,module,exports){
"use strict";
'use strict';
Object.defineProperty(exports, "__esModule", {value: true});
exports.string = undefined;
var _isnil = require('./lib/isnil.js');
var _string = require('./lib/string');
var isString = function isString(string) {
  return (0, _string._isString)(string);
};
var minLength = function minLength() {
  var min = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : Number.MIN_VALUE;
  return function(string) {
    return (0, _string._isString)(string) && string.length >= min;
  };
};
var maxLength = function maxLength() {
  var max = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : Number.MAX_VALUE;
  if ((0, _isnil.isNil)(max)) {
    max = Number.MAX_VALUE;
  }
  return function(string) {
    return (0, _string._isString)(string) && string.length <= max;
  };
};
var length = function length() {
  var min = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : Number.MIN_VALUE;
  var max = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : Number.MAX_VALUE;
  return function(string) {
    return minLength(min)(string) && maxLength(max)(string);
  };
};
var regex = function regex(expression) {
  return function(string) {
    var reg = new RegExp(expression);
    return reg.test(string);
  };
};
var string = exports.string = {
  isString: isString,
  minLength: minLength,
  maxLength: maxLength,
  length: length,
  regex: regex
};

},{"./lib/isnil.js":7,"./lib/string":11}],14:[function(require,module,exports){
"use strict";
'use strict';
Object.defineProperty(exports, "__esModule", {value: true});
exports.util = undefined;
var _array = require('./lib/array.js');
var _number = require('./lib/number.js');
var _object = require('./lib/object.js');
var _string = require('./lib/string.js');
var _string2 = require('./string.js');
var isRequired = function isRequired(value) {
  if ((0, _array._isArray)(value)) {
    return value.length > 0;
  }
  if ((0, _number._isNumber)(value)) {
    return value > 0;
  }
  if ((0, _object._isObject)(value)) {
    return Object.keys(value).length > 0;
  }
  if ((0, _string._isString)(value)) {
    return value.length > 0;
  }
  return Boolean(value);
};
var isEmail = function isEmail(email) {
  var sQtext = '[^\\x0d\\x22\\x5c\\x80-\\xff]';
  var sDtext = '[^\\x0d\\x5b-\\x5d\\x80-\\xff]';
  var sAtom = '[^\\x00-\\x20\\x22\\x28\\x29\\x2c\\x2e\\x3a-\\x3c\\x3e\\x40\\x5b-\\x5d\\x7f-\\xff]+';
  var sQuotedPair = '\\x5c[\\x00-\\x7f]';
  var sDomainLiteral = '\\x5b(' + sDtext + '|' + sQuotedPair + ')*\\x5d';
  var sQuotedString = '\\x22(' + sQtext + '|' + sQuotedPair + ')*\\x22';
  var sDomainRef = sAtom;
  var sSubDomain = '(' + sDomainRef + '|' + sDomainLiteral + ')';
  var sWord = '(' + sAtom + '|' + sQuotedString + ')';
  var sDomain = sSubDomain + '(\\x2e' + sSubDomain + ')*';
  var sLocalPart = sWord + '(\\x2e' + sWord + ')*';
  var sAddrSpec = sLocalPart + '\\x40' + sDomain;
  var sValidEmail = '^' + sAddrSpec + '$';
  return _string2.string.regex(sValidEmail)(email);
};
var isURL = function isURL(url) {
  var regex = /^(?:(?:https?|ftp):\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,}))\.?)(?::\d{2,5})?(?:[/?#]\S*)?$/i;
  return _string2.string.regex(regex)(url);
};
var isCep = function isCep(cep) {
  return _string2.string.regex(/^[0-9]{5}-[0-9]{3}$/)(cep);
};
var isHexColor = function isHexColor(color) {
  return _string2.string.regex(/^#([a-fA-F0-9]{6}|[a-fA-F0-9]{3})$/)(color);
};
var isCpf = function isCpf(cpfNumber) {
  var cpfRegex = /^(\d{3}[\.]?\d{3}[\.]?\d{3}[-]?\d{2})$/;
  if (cpfRegex.test(cpfNumber)) {
    var cpf = cpfNumber.replace(/\.|-/g, '');
    if (cpf === '00000000000' || cpf === '11111111111' || cpf === '22222222222' || cpf === '33333333333' || cpf === '44444444444' || cpf === '55555555555' || cpf === '66666666666' || cpf === '77777777777' || cpf === '88888888888' || cpf === '99999999999') {
      return false;
    }
    var final = void 0;
    var sum = 0;
    for (var firstIndex = 0; firstIndex < 9; firstIndex++) {
      sum += parseInt(cpf.charAt(firstIndex), 10) * (10 - firstIndex);
    }
    final = 11 - sum % 11;
    if (final === 10 || final === 11) {
      final = 0;
    }
    if (final !== parseInt(cpf.charAt(9), 10)) {
      return false;
    }
    sum = 0;
    for (var secondIndex = 0; secondIndex < 10; secondIndex++) {
      sum += parseInt(cpf.charAt(secondIndex), 10) * (11 - secondIndex);
    }
    final = 11 - sum % 11;
    if (final === 10 || final === 11) {
      final = 0;
    }
    if (final !== parseInt(cpf.charAt(10), 10)) {
      return false;
    }
    return true;
  }
  return false;
};
var isCnpj = function isCnpj(cnpjNumber) {
  var cnpjRegex = /^(\d{2}[\.]?\d{3}[\.]?\d{3}[\/]?\d{4}[-]?\d{2})$/;
  if (cnpjRegex.test(cnpjNumber)) {
    var cnpj = cnpjNumber.replace(/\.|-|\//g, '');
    if (cnpj === '00000000000000' || cnpj === '11111111111111' || cnpj === '22222222222222' || cnpj === '33333333333333' || cnpj === '44444444444444' || cnpj === '55555555555555' || cnpj === '66666666666666' || cnpj === '77777777777777' || cnpj === '88888888888888' || cnpj === '99999999999999') {
      return false;
    }
    var sum = 0;
    var length = 12;
    var num = cnpj.substring(0, length);
    var digits = cnpj.substring(length);
    var pos = length - 7;
    for (var firstIndex = length; firstIndex >= 1; firstIndex--) {
      sum += num.charAt(length - firstIndex) * pos--;
      if (pos < 2) {
        pos = 9;
      }
    }
    var results = sum % 11 < 2 ? 0 : 11 - sum % 11;
    if (results !== parseInt(digits.charAt(0), 10)) {
      return false;
    }
    length += 1;
    num = cnpj.substring(0, length);
    sum = 0;
    pos = length - 7;
    for (var secondIndex = length; secondIndex >= 1; secondIndex--) {
      sum += num.charAt(length - secondIndex) * pos--;
      if (pos < 2) {
        pos = 9;
      }
    }
    results = sum % 11 < 2 ? 0 : 11 - sum % 11;
    if (results !== parseInt(digits.charAt(1), 10)) {
      return false;
    }
    return true;
  }
  return false;
};
var util = exports.util = {
  isRequired: isRequired,
  isEmail: isEmail,
  isCep: isCep,
  isURL: isURL,
  isHexColor: isHexColor,
  isCpf: isCpf,
  isCnpj: isCnpj
};

},{"./lib/array.js":5,"./lib/number.js":8,"./lib/object.js":9,"./lib/string.js":11,"./string.js":13}],15:[function(require,module,exports){
"use strict";
'use strict';
Object.defineProperty(exports, "__esModule", {value: true});
exports.isValid = exports.validate = undefined;
var _string = require('./string');
Object.keys(_string).forEach(function(key) {
  if (key === "default" || key === "__esModule")
    return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _string[key];
    }
  });
});
var _util = require('./util');
Object.keys(_util).forEach(function(key) {
  if (key === "default" || key === "__esModule")
    return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _util[key];
    }
  });
});
var _boolean = require('./boolean');
Object.keys(_boolean).forEach(function(key) {
  if (key === "default" || key === "__esModule")
    return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _boolean[key];
    }
  });
});
var _number = require('./number');
Object.keys(_number).forEach(function(key) {
  if (key === "default" || key === "__esModule")
    return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _number[key];
    }
  });
});
var _date = require('./date');
Object.keys(_date).forEach(function(key) {
  if (key === "default" || key === "__esModule")
    return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _date[key];
    }
  });
});
var _creditCard = require('./creditCard');
Object.keys(_creditCard).forEach(function(key) {
  if (key === "default" || key === "__esModule")
    return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _creditCard[key];
    }
  });
});
var _ip = require('./ip');
Object.keys(_ip).forEach(function(key) {
  if (key === "default" || key === "__esModule")
    return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _ip[key];
    }
  });
});
var _isnil = require('./lib/isnil.js');
var _array = require('./lib/array.js');
var _reduce = require('./lib/reduce.js');
function _toConsumableArray(arr) {
  if (Array.isArray(arr)) {
    for (var i = 0,
        arr2 = Array(arr.length); i < arr.length; i++) {
      arr2[i] = arr[i];
    }
    return arr2;
  } else {
    return Array.from(arr);
  }
}
var validate = exports.validate = function validate() {
  for (var _len = arguments.length,
      fn = Array(_len),
      _key = 0; _key < _len; _key++) {
    fn[_key] = arguments[_key];
  }
  return function(value) {
    if ((0, _isnil.isNil)(value)) {
      return false;
    }
    return (0, _reduce.reduce)(function(acc, x) {
      if ((0, _isnil.isNil)(x)) {
        throw new Error('The function for validation is null or undefined.');
      } else {
        return acc && x(value);
      }
    }, true, fn);
  };
};
var isValid = exports.isValid = function isValid(v) {
  var r = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  return validate.apply(undefined, _toConsumableArray((0, _array._isArray)(r) ? r : [r]))(v);
};

},{"./boolean":1,"./creditCard":2,"./date":3,"./ip":4,"./lib/array.js":5,"./lib/isnil.js":7,"./lib/reduce.js":10,"./number":12,"./string":13,"./util":14}]},{},[15])(15)
});
